File: apps/server/src/routes/plex-home.ts
/**
 * Plex Home Users Management Routes
 * 
 * Endpoints for viewing and managing playlists across Plex Home users
 */

import { Router, Request, Response, NextFunction } from 'express';
import { requireAuth } from '../middleware/auth';
import { createValidationError, createInternalError } from '../middleware/error-handler';
import { logger } from '../utils/logger';

const router = Router();

// All routes require authentication
router.use(requireAuth);

/**
 * GET /api/plex-home/users
 * Get all Plex Home users on the account
 */
router.get('/users', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const userId = req.session.userId!;
    const db = (req.dbService as any).db;

    // Get user's Plex token
    const user = db.prepare('SELECT plex_token FROM users WHERE id = ?').get(userId);
    if (!user?.plex_token) {
      return next(createValidationError('No Plex token found'));
    }

    // Get home users from Plex.tv API
    const response = await fetch('https://plex.tv/api/v2/home/users', {
      headers: {
        'X-Plex-Token': user.plex_token,
        'X-Plex-Client-Identifier': 'playlist-lab-server',
        'X-Plex-Product': 'Playlist Lab',
        'X-Plex-Platform': 'Web',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      const errorText = await response.text();
      logger.error('[Plex Home] Failed to fetch home users', { 
        status: response.status, 
        statusText: response.statusText,
        error: errorText 
      });
      return next(createInternalError('Failed to fetch Plex Home users'));
    }

    const data = await response.json() as any;
    logger.info('[Plex Home] Raw API response', { data: JSON.stringify(data).substring(0, 1000) });
    
    // The response might be an array directly or wrapped in an object
    let usersArray = Array.isArray(data) ? data : (data.users || data.User || []);
    
    const homeUsers = usersArray.map((u: any) => ({
      id: u.uuid || u.id, // Use uuid if available, fallback to id
      title: u.title,
      username: u.username || u.title,
      thumb: u.thumb,
      admin: u.admin === 1 || u.admin === '1' || u.admin === true,
      restricted: u.restricted === 1 || u.restricted === '1' || u.restricted === true,
      guest: u.guest === 1 || u.guest === '1' || u.guest === true,
    }));

    logger.info('[Plex Home] Found home users', { count: homeUsers.length, userId, users: homeUsers });

    res.json({ homeUsers });
  } catch (error: any) {
    logger.error('[Plex Home] Error fetching home users', { error: error.message, stack: error.stack });
    next(createInternalError('Failed to fetch Plex Home users'));
  }
});

/**
 * POST /api/plex-home/users/:homeUserId/switch
 * Switch to a home user and get their token
 */
router.post('/users/:homeUserId/switch', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { homeUserId } = req.params as Record<string, string>;
    const userId = req.session.userId!;
    const db = (req.dbService as any).db;

    // Get admin user's Plex token
    const user = db.prepare('SELECT plex_token FROM users WHERE id = ?').get(userId);
    if (!user?.plex_token) {
      return next(createValidationError('No Plex token found'));
    }

    // Switch to the home user
    const response = await fetch(`https://plex.tv/api/v2/home/users/${homeUserId}/switch`, {
      method: 'POST',
      headers: {
        'X-Plex-Token': user.plex_token,
        'X-Plex-Client-Identifier': 'playlist-lab-server',
        'X-Plex-Product': 'Playlist Lab',
        'X-Plex-Platform': 'Web',
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      logger.error('[Plex Home] Failed to switch user', { homeUserId, status: response.status });
      return next(createInternalError('Failed to switch to home user'));
    }

    const data = await response.json() as any;
    
    logger.info('[Plex Home] Switched to home user', { homeUserId, userId });

    res.json({ 
      authToken: data.authToken,
      user: {
        id: data.id,
        title: data.title,
        username: data.username || data.title,
        thumb: data.thumb,
      },
    });
  } catch (error: any) {
    logger.error('[Plex Home] Error switching user', { error: error.message, stack: error.stack });
    next(createInternalError('Failed to switch to home user'));
  }
});

/**
 * GET /api/plex-home/users/:homeUserId/playlists
 * Get playlists for a specific home user
 * 
 * Note: Plex Home managed users share the same server. We use the admin token
 * to access the server, but playlists are user-specific based on who created them.
 */
router.get('/users/:homeUserId/playlists', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { homeUserId } = req.params as Record<string, string>;
    const userId = req.session.userId!;
    const db = (req.dbService as any).db;

    // Get admin user's Plex token and server info
    const user = db.prepare('SELECT plex_token FROM users WHERE id = ?').get(userId);
    if (!user?.plex_token) {
      return next(createValidationError('No Plex token found'));
    }

    const serverRow = db.prepare('SELECT server_url, access_token FROM user_servers WHERE user_id = ? LIMIT 1').get(userId);
    if (!serverRow?.server_url) {
      return next(createValidationError('No Plex server configured'));
    }

    // For managed users, we need to switch to get their token for the Plex server
    logger.info('[Plex Home] Attempting to switch to home user', { homeUserId, userId });
    
    const switchResponse = await fetch(`https://plex.tv/api/v2/home/users/${homeUserId}/switch`, {
      method: 'POST',
      headers: {
        'X-Plex-Token': user.plex_token,
        'X-Plex-Client-Identifier': 'playlist-lab-server',
        'X-Plex-Product': 'Playlist Lab',
        'X-Plex-Platform': 'Web',
        'Accept': 'application/json',
      },
    });

    if (!switchResponse.ok) {
      const errorText = await switchResponse.text();
      logger.error('[Plex Home] Failed to switch user for playlists', { 
        homeUserId, 
        status: switchResponse.status,
        statusText: switchResponse.statusText,
        error: errorText
      });
      return next(createInternalError('Failed to access home user playlists'));
    }

    const switchData = await switchResponse.json() as any;
    
    logger.info('[Plex Home] Switch response data', { 
      homeUserId,
      username: switchData.username,
      hasAuthToken: !!switchData.authToken,
      accountId: switchData.id
    });
    
    // Try to use the home user's token first
    const { PlexClient, resolvePlexToken } = await import('../services/plex');
    let playlists: any[] = [];
    let usedAdminToken = false;
    
    try {
      // Attempt to use the home user's token
      const userPlexClient = new PlexClient(serverRow.server_url, switchData.authToken);
      const allPlaylists = await userPlexClient.getPlaylists();
      playlists = allPlaylists.filter((p: any) => p.playlistType === 'audio');
      logger.info('[Plex Home] Successfully fetched playlists with home user token', { 
        homeUserId,
        count: playlists.length 
      });
    } catch (tokenError: any) {
      // If home user token fails, fall back to admin token and show all playlists
      logger.warn('[Plex Home] Home user token failed, using admin token', { 
        homeUserId,
        error: tokenError.message 
      });
      usedAdminToken = true;
      
      const adminPlexClient = new PlexClient(serverRow.server_url, resolvePlexToken(user, serverRow));
      const allPlaylists = await adminPlexClient.getPlaylists();
      playlists = allPlaylists.filter((p: any) => p.playlistType === 'audio');
      
      logger.info('[Plex Home] Fetched all playlists with admin token', { 
        homeUserId,
        count: playlists.length,
        note: 'Showing all playlists - cannot filter by user'
      });
    }
    
    // Map playlists to include proper fields for the frontend
    const mappedPlaylists = playlists.map((p: any) => ({
      id: p.ratingKey,
      name: p.title,
      trackCount: p.leafCount || 0,
      duration: p.duration || 0,
      composite: p.composite,
    }));
    
    logger.info('[Plex Home] Returning playlists', { 
      homeUserId,
      username: switchData.username,
      count: mappedPlaylists.length,
      usedAdminToken,
      userId 
    });

    res.json({ playlists: mappedPlaylists });
  } catch (error: any) {
    logger.error('[Plex Home] Error fetching home user playlists', { error: error.message, stack: error.stack });
    next(createInternalError('Failed to fetch home user playlists'));
  }
});

/**
 * POST /api/plex-home/playlists/:playlistId/copy
 * Copy a playlist from one home user to another
 */
router.post('/playlists/:playlistId/copy', async (req: Request, res: Response, next: NextFunction) => {
  try {
    const { playlistId } = req.params as Record<string, string>;
    const { sourceHomeUserId, targetHomeUserId, newName } = req.body;
    const userId = req.session.userId!;
    const db = (req.dbService as any).db;

    if (!sourceHomeUserId) {
      return next(createValidationError('sourceHomeUserId is required'));
    }

    // Get admin user's Plex token and server info
    const user = db.prepare('SELECT plex_token FROM users WHERE id = ?').get(userId);
    if (!user?.plex_token) {
      return next(createValidationError('No Plex token found'));
    }

    const serverRow = db.prepare('SELECT server_url, library_id, server_client_id, access_token FROM user_servers WHERE user_id = ? LIMIT 1').get(userId);
    if (!serverRow?.server_url) {
      return next(createValidationError('No Plex server configured'));
    }

    const { PlexClient, resolvePlexToken } = await import('../services/plex');

    // Get source user's token and playlist
    const sourceResponse = await fetch(`https://plex.tv/api/v2/home/users/${sourceHomeUserId}/switch`, {
      method: 'POST',
      headers: {
        'X-Plex-Token': user.plex_token,
        'X-Plex-Client-Identifier': 'playlist-lab-server',
        'X-Plex-Product': 'Playlist Lab',
        'X-Plex-Platform': 'Web',
        'Accept': 'application/json',
      },
    });

    if (!sourceResponse.ok) {
      return next(createInternalError('Failed to access source user'));
    }

    const sourceData = await sourceResponse.json() as any;
    const sourcePlexClient = new PlexClient(serverRow.server_url, sourceData.authToken);

    // Get playlist details
    const playlist = await sourcePlexClient.getPlaylistDetails(playlistId);
    const tracks = await sourcePlexClient.getPlaylistTracks(playlistId);

    // Get target user's token (or use current user if not specified)
    let targetToken = resolvePlexToken(user, serverRow);
    if (targetHomeUserId && targetHomeUserId !== 'current') {
      const targetResponse = await fetch(`https://plex.tv/api/v2/home/users/${targetHomeUserId}/switch`, {
        method: 'POST',
        headers: {
          'X-Plex-Token': user.plex_token,
          'X-Plex-Client-Identifier': 'playlist-lab-server',
          'X-Plex-Product': 'Playlist Lab',
          'X-Plex-Platform': 'Web',
          'Accept': 'application/json',
        },
      });

      if (!targetResponse.ok) {
        return next(createInternalError('Failed to access target user'));
      }

      const targetData = await targetResponse.json() as any;
      targetToken = targetData.authToken;
    }

    const targetPlexClient = new PlexClient(serverRow.server_url, targetToken);

    // Create playlist for target user
    const trackUris = tracks.map((t: any) => 
      `server://${serverRow.server_client_id}/com.plexapp.plugins.library/library/metadata/${t.ratingKey}`
    );

    const libraryUri = `server://${serverRow.server_client_id}/com.plexapp.plugins.library/library/sections/${serverRow.library_id}`;
    const playlistName = newName || playlist.title;

    const newPlaylist = await targetPlexClient.createPlaylist(playlistName, libraryUri, trackUris);

    // Copy cover art if it exists
    if (playlist.composite) {
      try {
        const coverUrl = `${serverRow.server_url}${playlist.composite}?X-Plex-Token=${sourceData.authToken}`;
        await targetPlexClient.uploadPlaylistPoster(newPlaylist.ratingKey, coverUrl);
      } catch (coverError) {
        logger.warn('[Plex Home] Failed to copy cover art', { playlistId, error: coverError });
      }
    }

    logger.info('[Plex Home] Copied playlist', { 
      sourcePlaylistId: playlistId,
      newPlaylistId: newPlaylist.ratingKey,
      sourceHomeUserId,
      targetHomeUserId: targetHomeUserId || 'current',
      userId 
    });

    res.json({ 
      success: true,
      playlist: {
        id: newPlaylist.ratingKey,
        title: newPlaylist.title,
        trackCount: tracks.length,
      },
    });
  } catch (error: any) {
    logger.error('[Plex Home] Error copying playlist', { error: error.message, stack: error.stack });
    next(createInternalError('Failed to copy playlist'));
  }
});

export default router;

File: internal/handlers/plex_home.go
// Plex Home users, ported from routes/plex-home.ts: browse the other
// managed/Home users on the admin's Plex account and copy one of their
// playlists into the current user's own account. Surfaced as an "Import"
// source, matching where the React app moved this to
// (components/ImportFromPlexHome.tsx's own doc comment: the old standalone
// Plex Home Users page's per-user playlist *management* - reordering, cover
// upload, add/replace tracks - never made the move, only this copy-to-self
// action did, since that's playlist import rather than playlist
// management).
package handlers

import (
	"fmt"
	"log/slog"
	"net/http"
	"net/url"

	"github.com/go-chi/chi/v5"

	"github.com/drevilish/playlist-lab/internal/auth"
	"github.com/drevilish/playlist-lab/internal/db"
	"github.com/drevilish/playlist-lab/internal/services/notifications"
	"github.com/drevilish/playlist-lab/internal/services/plex"
)

func RegisterPlexHome(r chi.Router, mw *auth.Middleware, h *ImportHandler) {
	r.Group(func(r chi.Router) {
		r.Use(mw.RequireAuth)
		r.Get("/import/plex-home/users", h.plexHomeUsers)
		r.Get("/import/plex-home/users/{homeUserId}/playlists", h.plexHomeUserPlaylists)
		r.Post("/import/plex-home/playlists/{playlistId}/copy", h.plexHomeCopyPlaylist)
	})
}

func (h *ImportHandler) plexHomeUsers(w http.ResponseWriter, r *http.Request) {
	user := auth.CurrentUser(r)
	dbUser, err := db.GetUserByID(h.DB, user.ID)
	if err != nil || dbUser.PlexToken == "" {
		http.Error(w, "No Plex token found", http.StatusBadRequest)
		return
	}
	users, err := h.PlexAuth.GetHomeUsersDetailed(dbUser.PlexToken)
	if err != nil {
		http.Error(w, "Failed to fetch Plex Home users", http.StatusBadGateway)
		return
	}
	h.Tmpl.RenderPartial(w, "partials/plex_home_users.html", map[string]any{"Users": users})
}

func (h *ImportHandler) plexHomeUserPlaylists(w http.ResponseWriter, r *http.Request) {
	user := auth.CurrentUser(r)
	homeUserID := chi.URLParam(r, "homeUserId")

	dbUser, err := db.GetUserByID(h.DB, user.ID)
	if err != nil || dbUser.PlexToken == "" {
		http.Error(w, "No Plex token found", http.StatusBadRequest)
		return
	}
	userServer, err := db.GetUserServer(h.DB, user.ID)
	if err != nil || userServer == nil {
		http.Error(w, "No Plex server configured", http.StatusBadRequest)
		return
	}

	// Prefer the home user's own token so only playlists they created show
	// up; fall back to the admin token (showing every playlist on the
	// server, unfiltered) if the switch fails - same fallback
	// routes/plex-home.ts's GET /users/:id/playlists uses, and for the same
	// reason: a managed (non-guest) user's token doesn't always carry
	// separate library permissions worth failing the whole request over.
	token, err := h.PlexAuth.SwitchToManagedUser(dbUser.PlexToken, homeUserID)
	usedAdminToken := false
	if err != nil {
		token = plex.ResolveToken(dbUser.PlexToken, userServer.AccessToken.String)
		usedAdminToken = true
	}

	client := plex.NewClient(userServer.ServerURL, token, h.PlexAuth.ClientID, "Playlist Lab")
	playlists, err := client.GetPlaylists()
	if err != nil {
		http.Error(w, "Failed to fetch home user playlists", http.StatusBadGateway)
		return
	}

	type playlistOption struct {
		ID, Name       string
		TrackCount     int
		SourceHomeUser string
	}
	rows := make([]playlistOption, 0, len(playlists))
	for _, p := range playlists {
		if p.PlaylistType != "audio" {
			continue
		}
		rows = append(rows, playlistOption{ID: p.RatingKey, Name: cleanPlaylistName(p.Title), TrackCount: p.LeafCount, SourceHomeUser: homeUserID})
	}

	h.Tmpl.RenderPartial(w, "partials/plex_home_playlists.html", map[string]any{
		"Playlists": rows, "UsedAdminToken": usedAdminToken, "HomeUserID": homeUserID,
	})
}

// plexHomeCopyPlaylist ports POST /api/plex-home/playlists/:playlistId/copy,
// scoped to this app's only real caller of it (ImportFromPlexHome.tsx):
// always copying into the current user's own account, never to a third
// home user. The TS route's targetHomeUserId branch existed for a
// standalone Plex Home management page that itself was never carried over
// to this rewrite (per this file's own package doc), so there is no
// caller left that ever sets it to anything but "current".
func (h *ImportHandler) plexHomeCopyPlaylist(w http.ResponseWriter, r *http.Request) {
	user := auth.CurrentUser(r)
	playlistID := chi.URLParam(r, "playlistId")
	_ = r.ParseForm()
	sourceHomeUserID := r.FormValue("sourceHomeUserId")
	newName := r.FormValue("newName")
	if sourceHomeUserID == "" {
		http.Error(w, "sourceHomeUserId is required", http.StatusBadRequest)
		return
	}

	dbUser, err := db.GetUserByID(h.DB, user.ID)
	if err != nil || dbUser.PlexToken == "" {
		http.Error(w, "No Plex token found", http.StatusBadRequest)
		return
	}
	userServer, err := db.GetUserServer(h.DB, user.ID)
	if err != nil || userServer == nil || !userServer.LibraryID.Valid {
		http.Error(w, "No Plex server/library configured", http.StatusBadRequest)
		return
	}

	sourceToken, err := h.PlexAuth.SwitchToManagedUser(dbUser.PlexToken, sourceHomeUserID)
	if err != nil {
		http.Error(w, "Failed to access source user", http.StatusBadGateway)
		return
	}
	sourceClient := plex.NewClient(userServer.ServerURL, sourceToken, h.PlexAuth.ClientID, "Playlist Lab")

	tracks, err := sourceClient.GetPlaylistTracks(playlistID)
	if err != nil {
		http.Error(w, "Failed to load source playlist", http.StatusBadGateway)
		return
	}
	playlists, err := sourceClient.GetPlaylists()
	if err != nil {
		http.Error(w, "Failed to load source playlist", http.StatusBadGateway)
		return
	}
	var sourceTitle, sourceComposite string
	for _, p := range playlists {
		if p.RatingKey == playlistID {
			sourceTitle, sourceComposite = p.Title, p.Composite
			break
		}
	}
	if newName == "" {
		newName = sourceTitle
	}

	machineID, err := sourceClient.GetMachineIdentifier()
	if err != nil {
		http.Error(w, "Failed to reach Plex server", http.StatusBadGateway)
		return
	}
	trackURIs := make([]string, 0, len(tracks))
	for _, t := range tracks {
		if t.RatingKey != "" {
			trackURIs = append(trackURIs, sourceClient.BuildTrackURI(t.RatingKey, machineID))
		}
	}

	targetToken := plex.ResolveToken(dbUser.PlexToken, userServer.AccessToken.String)
	targetClient := plex.NewClient(userServer.ServerURL, targetToken, h.PlexAuth.ClientID, "Playlist Lab")
	libraryURI := targetClient.BuildLibraryURI(userServer.LibraryID.String, machineID)
	newPlaylist, err := targetClient.CreatePlaylist(newName, libraryURI, trackURIs)
	if err != nil {
		http.Error(w, "Failed to create playlist", http.StatusBadGateway)
		return
	}

	if sourceComposite != "" {
		// The download request authenticates via the token embedded in the
		// URL itself (UploadPlaylistPoster attaches no header of its own),
		// same as plex-home.ts's own `${serverUrl}${composite}?X-Plex-Token=...`.
		coverURL := userServer.ServerURL + sourceComposite + "?X-Plex-Token=" + url.QueryEscape(sourceToken)
		if err := targetClient.UploadPlaylistPoster(newPlaylist.RatingKey, coverURL); err != nil {
			// Non-fatal, matching plex-home.ts's own try/catch around cover
			// copying: the playlist itself already exists with its tracks.
			slog.Warn("plex-home copy: failed to copy cover art", "error", err, "playlistId", playlistID)
		}
	}

	progress := 100
	h.Notifications.Add(user.ID, notifications.TypeImport, fmt.Sprintf("Imported from Plex Home: %s", newName), "Added", notifications.StatusSuccess, &progress)
	h.Tmpl.RenderPartial(w, "partials/notifications.html", map[string]any{"Notifications": h.Notifications.List(user.ID)})
}
